### DISCUSSION TITLE
How to speed up the Creation and Deletion of many Source nodes in RV ?

### DISCUSSION THREAD

**Discussion by mariapanneerrajan-spi** [AcademySoftwareFoundation/OpenRV#488] [Q&A] [answered]
We are working on creating an RV package that creates a Sequence-view of all the latest dailies of respective shots in a particular sequence of a show.

For this we need the ability in RV to create and delete multiple Source nodes based on the number of shots in a given sequence and connect them to an RVSequenceGroup node. And immediately after they are created, we need to get their properties such as cut_in, cut_out, scale, resolution, rotation,  etc.

The following is how we are added the Source groups,
```
source_groups = []
for media_path in media_paths:
    source_node = rv.commands.addSourceVerbose()
    source_group = rv.commands.nodeGroup(source_node)
    source_groups.append(source_group)
```
If we have 50 media_paths it takes about 9 secs to create the needed Source nodes.

And this is how we get the properties of the created source nodes,
```
for source_group in source_groups:
    flip = rv.commands.getIntProperty(f"{source_group}_transform2D.transform.flip")[0]
    flop = rv.commands.getIntProperty(f"{source_group}_transform2D.transform.flop")[0]
    scale = rv.commands.getIntProperty(f"{source_group}_transform2D.transform.scale")[0]    
    media_info  = rv.commands.sourceMediaInfo(f"{source_group}_source")
    start_frame = media_info.get("startFrame")
    end_frame = media_info.get("endFrame")
```

And the following is how we are deleting the nodes,
```
for source_group in source_groups:
    rv.commands.deleteNode(source_group)
```
And to delete 50 source nodes it takes about 18 seconds.

As more nodes get added to the current RV session, the time it takes to create and delete nodes increases, to a point where it becomes too slow to use.

We did try the progress loading technique using the `-progressiveSourceLoading 1` as mentioned in the following open-rv doc,
https://aswf-openrv.readthedocs.io/en/latest/rv-manuals/rv-reference-manual/rv-reference-manual-chapter-two.html#progressive-source-loading
While yes, the initial creation time of each of the source nodes is faster, since we need the ability to also fetch each of the source node's properties immediately as soon as they are respectively create, we get slowed down there.

Can you kindly provide us with some pointers on how we can speed up the Creation and Deletion of multiple Source nodes in RV please? Thank you.

Following the configuration of the machine in which we are getting these times,
CPU: Intel Xeon CPU E5-2620 v3 @ 2.40GHz x86_64
RAM: 62GB
OS: CentOS Linux release 7.6.1810 (Core)
OpenRv Version: 1.0.0 (d67263a)

**Comment by bernie-laberge** [comment, marked as answer]
Hello @mariapanneerrajan-spi ,
With the aim of speeding up adding multiple sources, I would recommend the two following RV commands:
addSourceBegin(), addSourceEnd():
"Optional call providing a fast add source mechanism when adding multiple sources which postpones connecting the added sources to the default views’ inputs until after the corresponding addSourceEnd() is called. The way to enable this optimization is to call addSourceBegin() first, followed by a bunch of addSource() calls, and then end with addSourceEnd()."

https://aswf-openrv.readthedocs.io/en/latest/rv-manuals/rv-reference-manual/rv-reference-manual-chapter-two.html#source-group-node

I do not recommend using the progressiveSourceLoading feature as it significantly increases the complexity of scripting because when it is enabled, a lot of RV commands become asynchronous.

As for optimizing the delete of these sources, I would try switching the view node to some other unrelated node using setViewNode() prior to deleting those sources as it might optimize the deletion. I suspect that the slow down comes from the fact that those nodes are displayed, which might generate lots of notifications when deleting those displayed nodes. But some additional profiling might be required to figure out what exactly is slowing down the deletion of those nodes.

I hope this helps.

**Comment by mariapanneerrajan-spi** [comment]
Hello Bernard,

Thanks for the prompt response.

Yes the pointers you have given speeds up the loading and deleting of the source nodes.

Thank you.
